src/main/java/hudson/remoting/Engine.java

```diff
@@ -496,6 +517,116 @@ public void run() {
+                            build(new Transport(session)));
+                    } catch (IOException x) {
+                        events.error(x);
+                    }
+                }
+                private void onMessage(byte[] message) {
+                    LOGGER.finest(() -> "received message of length " + message.length);
+                    receiver.handle(message);
+                }
+                @Override
+                public void onClose(Session session, CloseReason closeReason) {
+                    LOGGER.fine(() -> "onClose: " + closeReason);
+                    receiver.terminate(new ChannelClosedException(ch.get(), null));
+                }
+                @Override
+                public void onError(Session session, Throwable x) {
+                    LOGGER.log(Level.FINE, null, x);
+                    receiver.terminate(new ChannelClosedException(ch.get(), x));
+                }
+                class Transport extends AbstractByteArrayCommandTransport {
+                    final Session session;
+                    Transport(Session session) {
+                        this.session = session;
+                    }
+                    @Override
+                    public void setup(AbstractByteArrayCommandTransport.ByteArrayReceiver _receiver) {
+                        events.status("Setting up channel");
+                        receiver = _receiver;
+                    }
+                    @Override
+                    public void writeBlock(Channel channel, byte[] payload) throws IOException {
+                        LOGGER.finest(() -> "sending message of length " + payload.length);
+                        session.getBasicRemote().sendBinary(ByteBuffer.wrap(payload));
+                    }
+                    @Override
+                    public Capability getRemoteCapability() throws IOException {
+                        return headerHandler.remoteCapability;
+                    }
+                    @Override
+                    public void closeWrite() throws IOException {
+                        events.status("Write side closed");
+                        session.close();
+                    }
+                    @Override
+                    public void closeRead() throws IOException {
+                        events.status("Read side closed");
+                        session.close();
+                    }
+                }
+            }
+            ContainerProvider.getWebSocketContainer().connectToServer(new AgentEndpoint(),
+                ClientEndpointConfig.Builder.create().configurator(headerHandler).build(), URI.create(candidateUrls.get(0).toString().replaceFirst("^http", "ws") + "wsagents/"));
+            while (ch.get() == null) {
+                Thread.sleep(100);
+            }
+            LOGGER.info("Connected");
+            ch.get().join();
+            // TODO handle multiple candidate URLs
+            // TODO handle reconnection
```

Comment:
> Do we need to resolve these TODOs before release?

TBD. It is not like we would be introducing a regression; WebSocket mode is a new feature you need to opt in to, and for some users it is a choice between that and not connecting a certain kind of agent at all. Of course it would be good to implement these things.

Reply:
Yeah, I can make an argument either way. I'm concerned about introducing a new mode that is likely to cause people problems and report bugs. But, until people try it it's hard to know what the problems might be.